### ISSUE TITLE
Pnpm stuck in windows due to long file path

### ISSUE THREAD

**Issue by turtleinspace1** [pnpm/pnpm#7355]
### Verify latest release

- [ ] I verified that the issue exists in the latest pnpm release

### pnpm version

8.11.0

### Reproduction steps

Create a web project in windows
In Package.json look like follow, only install @angular/material
"dependencies": {
        "@angular/material": "16.1.3"
    },
Pnpm install
Stuck for 6 minutes

### Describe the Bug

Pnpm install need a so long time.
After debug the pnpm, the block happen on the follow code( fs.renameSync(oldPath, newPath);)
Pnpm will try to rename the folder for each dependency.
e.g. from 
D:\\.....s\\.pnpm\\registry.npmmirror.com+@angular+material@16.1.3_@angular+animations@17.0.4_@angular+cdk@16.1._dbljk25fedifsmmc22b3hvapj4\\node_modules\\@angular\\material_tmp_48644
to 
D:\\.....\\.pnpm\\registry.npmmirror.com+@angular+material@16.1.3_@angular+animations@17.0.4_@angular+cdk@16.1._dbljk25fedifsmmc22b3hvapj4\\node_modules\\@angular\\material
The path is so long exceed the limit of Max path in windows.
e.g.
D:\....\mui\node_modules\.pnpm\@angular+material@16.1.3_@angular+animations@17.0.4_@angular+cdk@16.1.3_@angular+common@17.0._p6gpzivryg5zt4x4wcyv73polm\node_modules\@angular\material\esm2022\core\selection\pseudo-checkbox
which cause the rename operation  stuck

### Expected Behavior
For whole project
pnpm install cost 50 minutes, it is too long
npm install will only cost about 6 minutes.

### Which Node.js version are you using?

18.17.1

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux

**Comment by ckolumbus** [comment]
same issue here

**Comment by zkochan** [comment]
I guess we'll optimize it in the next major version of pnpm. I have a PR open already: https://github.com/pnpm/pnpm/pull/7317

However, I don't understand why you get the `registry.npmmirror.com+` part in the path. Can you share your lockfile?

**Comment by turtleinspace1** [comment]
Thanks, it is very helpful, hope the fix 7317 can be in next major version, pnpm is excellent, but our project can not use the tool due to this problem.
For the registry.npmmirror.com+ part, I also feel weird .  but now I can not reproduce it.  my current version is  8.12.0
if it happen again, I will provide the lock files.

**Comment by knames** [comment]
also having this issue, spent 2 days trying to figure out why our project worked fine on my mac and not on my coworkers machines. Turns out our file names are long, and pnpm makes crazy nested path lengths. Even with long file paths turned on, it would still blow up.

**Comment by turtleinspace1** [comment]
I find the https://github.com/pnpm/pnpm/pull/7317 is closed and will not be integrated to V9 
May I know the reason？
When I use this solution, it can indeed solve my problem.

**Comment by zkochan** [comment]
The issue is still not solved for all users by that fix but it makes the directories in `.pnpm` less readable. The solution should be to allow long paths on Windows on system level. We should try to fix it with long path prefix:  https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry

**Comment by turtleinspace1** [comment]
Thanks for advice, I remember that my collogues already try to enable the long path. but it still could not work. But I will try again.
I guess it might be relevant to one of security policy on our develop environment. Some companies have their own IT department, and they always add some security policy to develop computer.

When I change the code and set MAX_LENGTH_WITHOUT_HASH  to lower value, then the problem is gone.

Maybe it could provide a choice in user side by extra parameter on the pnpm install? 
User can decide to use full path or use short path.

**Comment by zkochan** [comment]
To what value do you change MAX_LENGTH_WITHOUT_HASH?

**Comment by turtleinspace1** [comment]
> To what value do you change MAX_LENGTH_WITHOUT_HASH?
Sorry for later reply.
If  MAX_LENGTH_WITHOUT_HASH = 26 it worked
If MAX_LENGTH_WITHOUT_HASH  = 60, it also worked

The bigger value, I did not have a try.
Thanks for solving

**Comment by zkochan** [comment]
:ship: 9.1.0-0

**Comment by turtleinspace1** [comment]
I download the 9.1.0-0 and set virtual-store-dir-max-length=80 in .npmrc
Now it works normally, Thanks!

**Comment by Soltus** [comment]
Awesome, after upgrading to [v9.1.1](https://github.com/pnpm/pnpm/releases/tag/v9.1.1) `pnpm install` is hundreds of times faster in the workspace

**Comment by CallMeLaNN** [comment]
On the latest v9.13.2 I'm still having this issue on Windows 11 even after enabling the long path.

**I got it working** only by setting `virtual-store-dir-max-length=80` in `.npmrc` as https://github.com/pnpm/pnpm/issues/7355#issuecomment-2084495283 above.

The `pnpm install` is ok and fast but after that `pnpm astro check` throw import error:

```log
SyntaxError: Named export 'svelteInspector' not found. The requested module '@sveltejs/vite-plugin-svelte-inspector' is a CommonJS module, which may not support all module.exports as named exports.
```

I just realized this is the long path issue with pnpm running on a self-hosted github Action on windows since cloning directly on the same server is ok. The Action produce
`d:\actions-runner\_work\repo-name-abcd\repo-name-abcd\my-path\<actual repo>` **double folder** in the repo name (unlucky to have that long name) so `node_modules/.pnpm` is way longer.

Not sure if this is worth creating another issue...
